Issue #2530: Reset windowed lists

hawkfish opened the pull request:
Fix small issue with list quantiles.
Add pragmas for testing windowed aggregation algorithms
without recompilation

Commit:
Issue #2530: Reset windowed lists

Fix small issue with list quantiles.
Add pragmas for testing windowed aggregation algorithms
without recompilation

Mytherin commented on src/execution/operator/aggregate/physical_window.cpp:
```diff
@@ -390,7 +392,7 @@ static idx_t FindPrevStart(const BitArray<W> &mask, const idx_t l, idx_t r) {
 }
 
-static void MaterializeExpressions(Expression **exprs, idx_t expr_count, ChunkCollection &input,
+static void MaterializeExpressions(Expression **exprs, idx_t expr_count, ChunkCollection &input, // NOLINT
```
Are all these NOLINT's required?

hawkfish replied:
Well my editor complains (via our clangd settings) that there are two arguments that can be confused. I don't know why clang-tidy lets them through.

Mytherin replied:
Perhaps a different version of clang-tidy? Do you know what rule it is that gets triggered? This function seems rather innocent, perhaps we can just disable the offending rule.

hawkfish replied:
The rule is [bugprone-easily-swappable-parameters](https://clang.llvm.org/extra//clang-tidy/checks/bugprone-easily-swappable-parameters.html) I can disable it and remove the `NOLINT` flags.

Mytherin commented on src/include/duckdb/execution/window_segment_tree.hpp:
```diff
@@ -61,6 +62,12 @@ class WindowSegmentTree {
 	ChunkCollection *input_ref;
 
+	//! Use the window API, if available
+	bool use_window_api;
```
Perhaps store only the `WindowAggregationMode` and turn these into functions (`UseWindowAPI()`, `UseCombineAPI()`)?

hawkfish replied:
Sure. I did it this way because my temporary hack used the Booleans.

Mytherin commented on src/function/pragma/pragma_functions.cpp:
```diff
@@ -277,6 +277,20 @@ static void PragmaDebugManyFreeListBlocks(ClientContext &context, const Function
 	config.debug_many_free_list_blocks = true;
 }
 
+static void PragmaDebugWindowMode(ClientContext &context, const FunctionParameters &parameters) {
```
Could you add some tests where this is used, perhaps run some window functions in a loop toggling each of these different modes?

hawkfish replied:
Good point. And I think that should be part of the test suite for such functions to make sure they all give the same answer!

Mytherin reviewed (commented):
Thanks for the PR! Looks good. Some comments:

Commit:
Merge branch 'master' into hawkfish-window-reset

Commit:
Issue 2530: Reset windowed lists

Tweak window mode API.
Add implementation validation tests
for all window API aggregates.

Commit:
Issue #2530: Reset windowed lists

Suppress bugprone-easily-swappable-parameters in clang-tidy.

Commit:
Merge branch 'master' into hawkfish-window-reset

Mytherin commented:
Thanks for the updates!
